enterprise/server/remote_execution/uffd/uffd.go

```diff
@@ -17,9 +122,199 @@ func NewHandler() (*Handler, error) {
+	}
+
+	for {
+		// Poll UFFD for messages
+		_, pollErr := unix.Poll(pollFDs, -1)
+		if pollErr != nil {
+			if pollErr == unix.EINTR {
+				// Poll call was interrupted by another signal - retry
+				continue
+			}
+			return status.WrapError(pollErr, "poll uffd")
+		}
+
+		// Receive a page fault notification
+		guestFaultingAddr, err := getFaultingAddress(uffd)
+		if err != nil {
+			return err
+		}
+		guestPageAddr := pageStartAddress(guestFaultingAddr, pageSize)
+
+		// Find the memory data in the store that should be used to handle the page fault
+		faultStoreOffset, err := guestMemoryAddrToStoreOffset(guestPageAddr, mappings)
+		if err != nil {
+			return status.WrapError(err, "translate to store offset")
+		}
+		hostPageAddr, err := memoryStore.GetPageAddress(uintptr(faultStoreOffset), false /*=write*/)
+		if err != nil {
+			return status.WrapError(err, "get backing page address")
+		}
+
+		// Should never map a partial page at the end of the file, but just
+		// warn in this case for now.
+		if remainder := storeLength - int64(faultStoreOffset); remainder < int64(pageSize) {
+			log.CtxWarningf(ctx, "uffdio_copy range extends past store length")
+		}
+
+		// TODO(Maggie): Copy entire chunk, as opposed to just a page
+		_, err = resolvePageFault(uffd, uint64(guestPageAddr), uint64(hostPageAddr), uint64(pageSize))
+		if err != nil {
+			return err
+		}
+
+		log.Debugf("UFFDIO_COPY completed successfully")
+	}
+}
+
+// resolvePageFault copies `size` bytes of memory from a `Src` address to the faulting region `Dst`
+//
+// When complete, the UFFDIO_COPY call wakes up the process that triggered the page fault (When a process
+// attempts to access unallocated memory, it triggers a page fault and hangs until it has been resolved)
+//
+// Returns the number of bytes copied
+func resolvePageFault(uffd uintptr, faultingRegion uint64, src uint64, size uint64) (int64, error) {
+	copyData := uffdioCopy{
+		Dst: faultingRegion,
+		Src: src,
+		Len: uint64(size),
+	}
+	_, _, errno := syscall.Syscall(syscall.SYS_IOCTL, uffd, UFFDIO_COPY, uintptr(unsafe.Pointer(&copyData)))
+	if errno != 0 {
+		if errno == unix.ENOENT {
+			// The faulting process changed its virtual memory layout simultaneously with an outstanding UFFDIO_COPY
+			// The UFFDIO_COPY will have aborted and the handler should continue to serve page faults
+			return 0, nil
+		}
+		return 0, status.InternalErrorf("UFFDIO_COPY failed with errno(%d)", errno)
+	}
+	return int64(copyData.Copy), nil
+}
+
+// getFaultingAddress reads a notification from the uffd object and returns the faulting address
+// (i.e. the memory location the VM tried to access that triggered the page fault)
+func getFaultingAddress(uffd uintptr) (uint64, error) {
+	var event uffdMsg
+	_, _, errno := syscall.Syscall(syscall.SYS_READ, uffd, uintptr(unsafe.Pointer(&event)), unsafe.Sizeof(event))
+	if errno != 0 {
+		return 0, status.InternalErrorf("read event from uffd failed with errno(%d)", errno)
```

Comment:
nit: logging the stringified errno (like EINVAL etc) might make it slightly easier to debug

```suggestion
		return 0, status.InternalErrorf("read event from uffd failed with errno %d: %s", errno, errno)
```

Reply:
Unfortunately there's no way to get the symbolic name (like EINVAL) of an errno. Printing the stringified errno will print a human readable string.

Ex. For errno(2) ENOENT - it would print "No such file or directory"

I found this confusing to debug because ENOENT means something very different in the context of UFFD.